media: i2c: addi9036: Add ADDI9036 driver

btogorean opened the pull request:
Initial version of V4L driver for ADDI9036 CCD TOF Signal Processor.

Signed-off-by: Bogdan Togorean <bogdan.togorean@analog.com>

mirceacaprioru commented on Documentation/devicetree/bindings/media/i2c/addi9036.yaml:
```diff
@@ -0,0 +1,57 @@
+
+  port:
+    type: object
+    description:
```
i think you need the | for multiline comment

btogorean replied:
ack

mirceacaprioru commented on Documentation/devicetree/bindings/media/i2c/addi9036.yaml:
```diff
@@ -0,0 +1,57 @@
+# Copyright 2019 Analog Devices Inc.
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/bindings/media/i2c/addi9036.yaml#
```
should it be adi,addi9036 ?

btogorean replied:
ack

mirceacaprioru commented on Documentation/devicetree/bindings/media/i2c/addi9036.yaml:
```diff
@@ -0,0 +1,57 @@
+
+examples:
+  - |
+
```
nitpick extra line

btogorean replied:
ack

mirceacaprioru commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,722 @@
+
+			ret = addi9036_set_power_off(addi9036);
+			if (ret < 0) {
+				mutex_unlock(&addi9036_lock);
```
can be moved to exit

btogorean replied:
global mutex will be removed

commodo commented on Documentation/devicetree/bindings/media/i2c/addi9036.yaml:
```diff
@@ -0,0 +1,57 @@
+                        data-lanes = <1 2>;
+                };
+        };
+    };
```
usually, the dt-binding doc is a commit after the driver commit

btogorean replied:
ack

commodo commented on Documentation/devicetree/bindings/media/i2c/addi9036.yaml:
```diff
@@ -0,0 +1,57 @@
+examples:
+  - |
+
+    addi9603_camera@64 {
```
this looks like a typo: `addi9603_camera` -> `addi9036_camera`

btogorean replied:
ack, thx

commodo commented on Documentation/devicetree/bindings/media/i2c/addi9036.yaml:
```diff
@@ -0,0 +1,57 @@
+$id: http://devicetree.org/schemas/bindings/media/i2c/addi9036.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: Analog Devices CCD TOF Signal Processor
```
maybe add the part name in the title `Analog Devices ADDI9036 CCD TOF Signal Processor`

btogorean replied:
ack

commodo commented on drivers/media/i2c/Kconfig:
```diff
@@ -544,6 +544,18 @@ comment "Camera sensor devices"
+	depends on I2C && VIDEO_V4L2 && VIDEO_V4L2_SUBDEV_API
+	depends on MEDIA_CAMERA_SUPPORT
+	select V4L2_FWNODE
+	---help---
```
i think `help` is preferred over `---help---`  ; but not sure ; maybe check what upstream prefers these days

btogorean replied:
I checked an help is preferred. will modify, thx

commodo commented on include/uapi/linux/v4l2-controls.h:
```diff
@@ -933,6 +933,9 @@ enum v4l2_dv_it_content_type {
 
+/* The base for the addi9036 driver controls, 4 controls are reserved */
+#define V4L2_CID_USER_ADDI9036_BASE		(V4L2_CID_DV_CLASS_BASE + 0x100)
+
```
maybe this should be moved at line 191, closer to the other group of USER base classes

btogorean replied:
ack

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,722 @@
+ *
+ * Based on OV5640.c
+ *
+ * This program is free software; you can redistribute it and/or modify
```
this license text can be removed if adding `// SPDX-License-Identifier: GPL-2.0+`

btogorean replied:
ack

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,722 @@
+// SPDX-License-Identifier: GPL-2.0+
```
this needs to be `SPDX-License-Identifier: GPL-2.0`  because `MODULE_LICENSE("GPL v2");`
that plus seems to be important to upstream & some lawyers;

btogorean replied:
ack, thx

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,722 @@
+#include <linux/version.h>
+#include <media/v4l2-ctrls.h>
+#include <media/v4l2-fwnode.h>
+#include <media/v4l2-subdev.h>
```
curios: are all these headers required?
i.e. can this list be reduced?

btogorean replied:
linux/version.h and linux/regulator/consumer.h can be removed. thx

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,722 @@
+			}
+
+			mutex_unlock(&addi9036_lock);
+			dev_dbg(addi9036->dev, "power off complete\n");
```
can this debug print be removed ?

btogorean replied:
most of the dbg_prints are there just to help with the development of the higher level ADI ToF SDK. I'll check with the developers of the SDK and remove them if are not necessary any more.

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,722 @@
+#include <media/v4l2-fwnode.h>
+#include <media/v4l2-subdev.h>
+
+static DEFINE_MUTEX(addi9036_lock);
```
this global mutex seems to be used together with the power_lock;
this means that if you have more than 2 devices, this global lock won't allow both of them to do any power on/off separately;

is there a use-case where this is needed?

btogorean replied:
the global mutex is not useful any more. will be removed

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,722 @@
+	/* If the power count is modified from 0 to != 0 or from != 0 to 0,
+	 * update the power state.
+	 */
+	if (addi9036->power_count == !on) {
```
what is the `power_count` needed for?
the logic seems a bit more complicated than it needs to be;

would this work?

```
mutex_lock(&addi9036->power_lock);

if (addi9036->power_on == on)
   goto out_unlock;

if (on)
    ret = addi9036_set_power_on(addi9036);
else
    ret = addi9036_set_power_off(addi9036);

if (ret < 0)
    goto out_unlock;

addi9036->power_on = on;

out_unlock:
mutex_unlock(&addi9036->power_lock);
```

also, i don't understand the need for `mutex_lock(&addi9036_lock);` (the global lock);
maybe there is a use-case i am missing

btogorean replied:
all that power_count logic is to avoid calling again addi9036_set_power_on if the camera is already on.

commodo replied:
ok, but why a `counter` logic?
isn't a bool type sufficient?
incrementing & decrementing has it's own exposure to potential faults from outside elements

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,722 @@
+		reg += 2;
+		val += 2;
+	}
+	dev_dbg(addi9036->dev, "%s, total successful write count %d\n",
```
curios: is this debug print still needed?
can we remove it and others like it?

btogorean replied:
same as before

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,722 @@
+	val = ((unsigned short *)(ctrl->p_new.p_u16 + 1));
+
+	while (index < ctrl->elems) {
+		if (*reg) {
```
curious: why `if (*reg)` ?

btogorean replied:
there was a problem with SDK requests. It requested to write in register 0 with data 0 but that is not a valid register number. Will be removed

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,722 @@
+
+		ret = addi9036_set_chip_config(addi9036, ctrl);
+		break;
+
```
nitpick: extra line

btogorean replied:
ack

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,722 @@
+	struct addi9036 *addi9036 = to_addi9036(sd);
+
+	dev_dbg(addi9036->dev, "Mbus code index %d\n", code->index);
+	if (code->index > 0)
```
this looks weird;
curios: why return -EINVAL if (code->index > 0)?

btogorean replied:
To enumerate all media bus formats available applications initialize the index field of v4l2_subdev_mbus_code_enum structure starting with 0, in increments of 1 until -EINVAL is returned. In our case we support only one format, that is returned at index 0.

commodo replied:
ack

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,722 @@
+{
+	unsigned int i;
+
+	for (i = ARRAY_SIZE(addi9036_mode_info_data) - 1; i > 0; i--) {
```
this assumes there is a specific order in which `addi9036_mode_info_data` is organized;
than can be fine, but maybe add a comment to the definition of the `addi9036_mode_info_data` table to make sure future additions take this into consideration

btogorean replied:
ack. comment will be added to avoid future errors. thx

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,722 @@
+{
+	struct addi9036 *addi9036 = to_addi9036(sd);
+	struct v4l2_mbus_framefmt *__format;
+	struct v4l2_rect *__crop;
```
maybe rename `__crop` -> `crop` ; I don't see any collisions with another var-name
and `__format` -> `fmt` or  `__format` -> `framefmt` to avoid the collision

btogorean replied:
ack

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,722 @@
+	return 0;
+}
+
+static int addi9036_s_stream(struct v4l2_subdev *subdev, int enable)
```
this does not seem to do anything;

btogorean replied:
true but if .video from `v4l2_subdev_ops` is NULL `VIDIOC_DQBUF` ioctl will fail

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,722 @@
+		return PTR_ERR(addi9036->regmap);
+	}
+
+	endpoint = of_graph_get_next_endpoint(dev->of_node, NULL);
```
is there a fwnode get endpoint alternative?
to avoid using `of_graph` direclty?

btogorean replied:
will replace with `fwnode_graph_get_next_endpoint`. thx

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,722 @@
+	v4l2_async_unregister_subdev(&addi9036->sd);
+	media_entity_cleanup(&addi9036->sd.entity);
+	v4l2_ctrl_handler_free(&addi9036->ctrls);
+	mutex_destroy(&addi9036->power_lock);
```
`mutex_destroy()` are typically discouraged to do in `remove`;
since the mutex struct is static data, it should be free'd together with `addi9036`

btogorean replied:
ack

btogorean commented:
Changes in V2:

- fixed dt-bindings problems
- removed unnecessary included headers
- removed global mutex
- removed dev_dbg print functions
- removed check for reg == 0 on `addi9036_set_chip_config`
- nitpicks fixed

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,673 @@
+ * Based on OV5640 Camera Driver
+ * Copyright (C) 2011-2013 Freescale Semiconductor, Inc. All Rights Reserved.
+ * Copyright (C) 2014-2017 Mentor Graphics Inc.
+ *
```
nitpick: this line is empty;
also, maybe it's a good idea to move `Copyright (C) 2018-2019 Analog Devices, All Rights Reserved.` here, to keep things chronologically sorted;

also, I am confused why there is a copyright for Freescale and Mentor Graphics.
if they have some involvement in this driver, maybe it's worth mentioning in the commit;

btogorean replied:
Their involvment is in the OV5640 driver on which this driver is based on.

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,673 @@
+
+	while (index < ctrl->elems) {
+		ret = regmap_write(addi9036->regmap, *reg, *val);
+		if (!ret)
```
the `count` is incremented but unused;
remove it ?

btogorean replied:
will be removed. thx

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,673 @@
+		if (!ret)
+			count++;
+		else {
+			dev_info(addi9036->dev,
```
this sounds more like `dev_warn` or `dev_err` depending on the severity of the error

btogorean replied:
ack

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,673 @@
+	case V4L2_CID_LINK_FREQ:
+		break;
+	default:
+		dev_dbg(addi9036->dev, "%s > Unhandled: %x  param=%x\n",
```
this sounds like `dev_err` ?

btogorean replied:
ack

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,673 @@
+	reg = (unsigned short *)(ctrl->p_new.p_u16);
+	val = ((unsigned short *)(ctrl->p_new.p_u16 + 1));
+
+	while (index < ctrl->elems) {
```
this is fine to leave as is, but this does look like a for loop with increment 2;
so  `for (index = 0; index < ctrl->elems; index += 2)`

btogorean replied:
ack

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,673 @@
+	case V4L2_CID_AD_DEV_REG_READ_QUERY:
+		ret = regmap_read(addi9036->regmap, *(u16 *)(ctrl->p_new.p_u16),
+				  &val);
+		*(u16 *)(ctrl->p_new.p_u16) = val;
```
do we need to do err check if `ret != 0` ?

btogorean replied:
will add

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,673 @@
+		return &addi9036->fmt;
+	default:
+		/* Returning null causes 0 derefence in addi9036_get_format */
+		return &addi9036->fmt;
```
curios: so, why not return `NULL` and treat it as an error in `addi9036_get_format()` & `addi9036_set_format()`?

btogorean replied:
ack

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,673 @@
+	if (sel->target != V4L2_SEL_TGT_CROP)
+		return -EINVAL;
+
+	sel->r = *__addi9036_get_pad_crop(addi9036, cfg, sel->pad, sel->which);
```
style-wise: is there a need to prefix this function with `__`?
why not just `addi9036_get_pad_crop`?

btogorean replied:
was just a leftover. will rename

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,673 @@
+		addi9036->current_mode = new_mode;
+	}
+
+	framefmt = __addi9036_get_pad_format(addi9036, cfg, format->pad,
```
also, about this function: why not `__addi9036_get_pad_format`?

btogorean replied:
same as before

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,673 @@
+	const struct addi9036_mode_info *new_mode;
+	int ret;
+
+	dev_dbg(addi9036->dev, "set_fmt: %x %dx%d\n",
```
is this left here intentionally or forgotten?

btogorean replied:
left here intentionally. Is useful in sdk development

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,673 @@
+	30060000
+};
+
+/* Elements of the structure must be ordered ascending by width */
```
`/* Elements of the structure must be ordered ascending by width */`
the `addi9036_find_nearest_mode()` function seems to also consider the height;
so
`/* Elements of the structure must be ordered ascending by width & height */`
?

btogorean replied:
ack

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,673 @@
+	struct addi9036 *addi9036;
+	int ret;
+
+	dev_dbg(dev, "%s: i2c addr = 0x%x\n", __func__, client->addr);
```
is this intentionally left here or forgotten?

btogorean replied:
intentionally left

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,673 @@
+	if (addi9036->link_freq)
+		addi9036->link_freq->flags |= V4L2_CTRL_FLAG_READ_ONLY;
+
+
```
nitpick: extra line

btogorean replied:
ack. thx

commodo commented on include/uapi/linux/v4l2-controls.h:
```diff
@@ -190,6 +190,9 @@ enum v4l2_colorfx {
 #define V4L2_CID_USER_IMX_BASE			(V4L2_CID_USER_BASE + 0x1090)
 
+/* The base for the addi9036 driver controls, 4 controls are reserved */
+#define V4L2_CID_USER_ADDI9036_BASE		(V4L2_CID_DV_CLASS_BASE + 0x100)
```
curios: why `V4L2_CID_DV_CLASS_BASE` and not `V4L2_CID_USER_BASE`?

btogorean replied:
we thought that these controls fit better in Digital Video controls group. To be sincere I don't know if this is the best place but in order to change them the SDK needs also to be updated. So I propose to keep them as it is now.

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,673 @@
+
+static struct i2c_driver addi9036_i2c_driver = {
+	.driver			= {
+		.of_match_table = of_match_ptr(addi9036_of_match),
```
`of_match_ptr()` is not recommended;
this will cause problems when this driver is used with ACPI and not OF;
this is more recommended `.of_match_table = addi9036_of_match,`

btogorean replied:
ack

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,673 @@
+		goto free_entity;
+	}
+
+	addi9036_entity_init_cfg(&addi9036->sd, NULL);
```
I am not completely sure that passing NULL for `cfg` is safe down the call-stack.
so, curios: is this safe down the call-stack? do all functions handle the NULL case?

btogorean replied:
checked and is safe

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,673 @@
+	struct addi9036 *addi9036 = to_addi9036(subdev);
+	int ret = 0;
+
+	dev_dbg(addi9036->dev, "stream %d\n", enable);
```
intentionally left here or forgotten?

also, why define `ret = 0`?
why not just `return 0`?

btogorean replied:
intentionally left. useful for SDK development

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,673 @@
+	int count = 0;
+
+	reg = (unsigned short *)(ctrl->p_new.p_u16);
+	val = ((unsigned short *)(ctrl->p_new.p_u16 + 1));
```
nitpick: the extra parentheses seems un-needed

btogorean replied:
ack

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,673 @@
+	u32 width;
+	u32 height;
+	u32 pixel_clock;
+	u32 link_freq;
```
there are multiple definitions of `link_freq`; and it seems a bit confusing to find them; i see a table `link_freq` [and i forgot what else]
maybe rename this to `link_freq_idx` or something like that;

btogorean replied:
ack, will rename

btogorean replied:
ack

commodo commented on drivers/media/i2c/addi9036.c:
```diff
@@ -0,0 +1,673 @@
+	{ 0xc4f0, 0x0001 }
+};
+
+static const s64 link_freq[] = {
```
this could be `link_freq_tbl` or something like that; to make it clear that it's a table;

btogorean replied:
ack

btogorean commented:
Changes in V3:

- renamed the link_freq tables/indexes
- nitpicks fixed
- removed `of_match_ptr()`
- removed prefix `__`
- removed unused `count`
